migration/table.go

```diff
@@ -0,0 +1,249 @@
+	dropColumnFromTable = `ALTER TABLE %s DROP COLUMN IF EXISTS %v`
+	renameColumnInTable = `-- ALTER TABLE %s RENAME COLUMN %v TO %v; -- uncomment to activate, remove ADD/DROP COLUMN above and below` // Can't have IF EXISTS here
+
+	dropTable = `DROP TABLE IF EXISTS %s`
+)
+
+// TableCreator handles creation of schema.Table in database as SQL strings
+type TableCreator struct {
+	log     hclog.Logger
+	dialect schema.Dialect
+}
+
+func NewTableCreator(log hclog.Logger, dialect schema.Dialect) *TableCreator {
+	return &TableCreator{
+		log:     log,
+		dialect: dialect,
+	}
+}
+
+// CreateTable generates CREATE TABLE definitions for the given table and runs them on the given conn
+func (m TableCreator) CreateTable(ctx context.Context, conn schema.QueryExecer, t, p *schema.Table) error {
+	ups, _, err := m.CreateTableDefinitions(ctx, t, p)
+	if err != nil {
+		return err
+	}
+	for _, sql := range ups {
+		if err := conn.Exec(ctx, sql); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// CreateTableDefinitions reads schema.Table and builds the CREATE TABLE and DROP TABLE statements for it, also processing and returning subrelation tables
+func (m TableCreator) CreateTableDefinitions(ctx context.Context, t *schema.Table, parent *schema.Table) (up, down []string, err error) {
+	b := &strings.Builder{}
+
+	// Build a SQL to create a table
+	b.WriteString("CREATE TABLE IF NOT EXISTS " + strconv.Quote(t.Name) + " (\n")
+
+	for _, c := range m.dialect.Columns(t) {
+		b.WriteByte('\t')
+		b.WriteString(strconv.Quote(c.Name) + " " + m.dialect.DBTypeFromType(c.Type))
+		if c.CreationOptions.NotNull {
+			b.WriteString(" NOT NULL")
+		}
+		// c.CreationOptions.Unique is handled in the Constraints() call below
+		b.WriteString(",\n")
+	}
+
+	cons := m.dialect.Constraints(t, parent)
+	for i, cn := range cons {
+		b.WriteByte('\t')
+		b.WriteString(cn)
+
+		if i < len(cons)-1 {
+			b.WriteByte(',')
+		}
+
+		b.WriteByte('\n')
+	}
+
+	b.WriteString(")")
+
+	up, down = make([]string, 0, 1+len(t.Relations)), make([]string, 0, 1+len(t.Relations))
+	up = append(up, b.String())
+	up = append(up, m.dialect.Extra(t, parent)...)
+
+	// Create relation tables
+	for _, r := range t.Relations {
+		if cr, dr, err := m.CreateTableDefinitions(ctx, r, t); err != nil {
+			return nil, nil, err
+		} else {
+			up = append(up, cr...)
+			down = append(down, dr...)
+		}
+	}
+
+	down = append(down, fmt.Sprintf(dropTable, t.Name))
+
+	return up, down, nil
+}
+
+// UpgradeTable reads current table info from the given conn for the given table, and returns ALTER TABLE ADD COLUMN statements for the missing columns.
+// Newly appearing tables will return a CREATE TABLE statement
+// Column renamings are detected (best effort) and ALTER TABLE RENAME COLUMN statements are generated as comments.
+// Table renamings or removals are not detected.
+func (m TableCreator) UpgradeTable(ctx context.Context, conn *pgxpool.Conn, t, parent *schema.Table) (up, down []string, err error) {
```

Comment:
Doesn't this find the TableDiffs? rather than "Upgrade", Initially reading this I taught this upgrades the table but it GeneratesTablesDiffs

Reply:
Renamed to `DiffTable`